v-next/hardhat-viem/src/chains.ts

```diff
@@ -0,0 +1,139 @@
+import type { TestClientMode } from "./types.js";
+import type { EthereumProvider } from "@ignored/hardhat-vnext/types/providers";
+import type * as viemT from "viem";
+
+import { HardhatError } from "@ignored/hardhat-vnext-errors";
+import { extractChain } from "viem";
+import * as chainsModule from "viem/chains";
+import { hardhat, anvil } from "viem/chains";
+
+/* eslint-disable-next-line @typescript-eslint/consistent-type-assertions
+-- TODO: this assertion should not be necessary */
+const chains = Object.values(chainsModule) as viemT.Chain[];
```

Comment:
I had to cast this, but the [example](https://viem.sh/docs/utilities/extractChain.html#usage) in the Viem docs doesn’t require a cast. Not entirely sure why, though.

Reply:
This ensures a clear type for chainId to meet the requirements of the extractChain function.
```ts
export function checkChainId(chainId: number): (typeof chains)[number]["id"] {
	const findId = chains.find((chain) => chain.id === chainId);
	if (findId === undefined) {
		throw new HardhatError(HardhatError.ERRORS.VIEM.NETWORK_NOT_FOUND, {
			chainId,
		});
	}
	return findId.id;
}

export async function getChainId(
	provider: EthereumProvider,
): Promise<ReturnType<typeof checkChainId>> {
	const cachedChainId = chainIdCache.get(provider);
	if (cachedChainId !== undefined) {
		return checkChainId(cachedChainId);
	}

	const chainId = Number(await provider.request({ method: "eth_chainId" }));
	chainIdCache.set(provider, chainId);

	return checkChainId(chainId);
}

```

Reply:
Hi @iosh, thanks for your feedback! However, this PR is intended for internal review by the team, as there is specific context around v3 that external contributors might not have. For example, we need the `module` option to be `"node16"` to support ESM, so changing it to `"CommonJS"` wouldn't work. 

We’ll start accepting feedback once we have a more stable version of v3. In the meantime, feel free to contribute to or provide feedback on v2 issues. Look for issues not labeled as `v-next`.